nslook/nslook.1
.TH NSLOOK 1 "Sep 10, 1991"
.UC 4
.SH NAME
nslook \- name server lookup
.SH SYNOPSIS
.B nslook
[
.B \-l
]
[
.B \-n
]
.I hostname
.br
.B nslook
[
.B \-l
]
[
.B \-a
]
.I IP-address
.br
.SH DESCRIPTION
.B Nslook 
provides a simple way to retrieve an IP address or a host name
from the Domain Name Service. The command
.PP
.ti +10n
\fC nslook \fP
.I hostname
.PP
prints the IP address of
.I hostname
on standard output (and nothing else). The command
.PP
.ti +10n
\fC nslook \fP
.I IP-address
.PP
prints the official name of the Internet host with the specified
.I IP-address
on standard output.
.PP
If the
.B \-l
flag is given, 
.B nslook
gives a long format display, showing the host's official name, zero or
more aliases, and one or more IP addresses.
.B \-a
forces the interpretation of the argument as an address, 
.B \-n 
as a hostname. Otherwise it is treated as an address if it ``looks like'' 
an address, and as a hostname if not.
.PP
.SH DIAGNOSTICS
The diagnostics of 
.B nslook
correspond to the errors signalled by
gethostbyname(3) and gethostbyaddr(3).
.TP
\fC unknown host \fP
.TP
\fC temporary error -- try again later \fP
This usually indicates that the local server did not receive
a response from an authoritative server.
A retry at some later time may succeed.
.TP
\fC unrecoverable error \fP
Some unexpected server failure was encountered.
.TP
\fC no IP address \fP
The requested name is valid but does not have an IP address; 
this is not a temporary error.  
This means that the name is known to the name server but there is no address
associated with this name.
Another type of request to the name server using this domain name
will result in an answer;
for example, a mail-forwarder may be registered for this domain.
.PP
The exit code of 
.B nslook
in case of one these errors is the corresponding value of h_errno as defined
in /usr/include/netdb.h: HOST_NOT_FOUND, TRY_AGAIN, NO_RECOVERY, or NO_DATA
(NO_ADDRESS on some systems).
.PP
If an unknown error occurs, the exit code is 16.
In case of a usage error the exit code is 15.
.SH "SEE ALSO"
nslookup(1), gethostbyaddr(3), gethostbyname(3)
.SH FILES
/usr/include/netdb.h
.SH BUGS
Normally
.B nslook
does not list alias hostnames since it only uses the information given 
by gethostbyname(3) or gethostbyaddr(3).
.PP
If 
.B nslook
tries to find the address of a non-existent host in the local domain, 
the error ``no IP address'' is signalled. This is due to the return code 
of gethostbyname(3).
.PP
.SH AUTHOR
Juergen Nickelsen  <nickel@cs.tu-berlin.de>


rrc2/history
 ############################################################################
#                                                                            #
#                      History of my Implementation of                       #
#                      RRC.2 - Ron Rivest Cipher Nr. 2                       #
#                                                                            #
 ############################################################################

13.02.96:
	Received specification

15.02.96: 
	Linux (ANSI-) C version done;
	includes RRC.2 library, test program, and simple sample crypt
	program

18.02.96:
	First announcement on usenet (rrc2_news.txt)

18.02.96:
	Turbo Pascal version done;
	Includes only Unit RRC2 and test program.

20.02.96:
	Improved C version: now endian neutral (untested)

20.02.96:
	RRC.2 obtainable via ftp

23.02.96:
	second usenet announcement (rrc2_news_2.txt)

# version FINI


nslook/nslook.c
/*
 *  Name:              nslook.c
 *  Author:            nickel@cs.tu-berlin.de
 *  Version:           1.3 
 *  State:             save
 *  Last modification: Mon Sep 16 13:47:29 1991 by nickel@cs.tu-berlin.de
 */

/*
 * nslook.c  - simple Domain Name Service lookup. This is realized
 * just as a front end to gethostbyname() and gethostbyaddr().
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdio.h>

/* exit codes: */
#define USAGE_ERROR 15		/* distinguish this from possible */
				/* values in h_errno */
#define UNKNOWN_ERROR 16	/* because h_errno is zero sometimes */
				/* even when gethostbyname() returns NULL */

char *progname ;
extern struct hostent *gethostbyname(), *gethostbyaddr() ;
extern int errno ;

main(argc, argv)
int argc ;
char **argv ;
{
    extern int optind ;		/* from getopt */
    int opt ;			/* option character */
    int aflg = 2 ;		/* 0=name, 1=address, 2=guess */
    int lflg = 0 ;		/* long output format */
    int errflg = 0 ;		/* usage error encountered */
    char *cp, *strrchr() ;

    progname = argv[0] ;
    if (cp = strrchr(progname, '/')) progname = cp + 1 ;
    while ((opt = getopt(argc, argv, "vlan")) != -1) {
	switch (opt) {
	  case 'l':		/* long */
	  case 'v':		/* verbose */
	    lflg = 1 ;
	    break ;
	  case 'a':		/* argument is address */
	    aflg = 1 ;
	    break ;
	  case 'n':		/* argument is name */
	    aflg = 0 ;
	    break ;
	  case '?':		/* give usage message */
	    errflg++ ;
	}
    }
    if (errflg || optind != argc - 1) {
	usage() ;
	exit(USAGE_ERROR) ;	/* distinguish from the other errors */
    }

    if (aflg == 2) {
      int i;			/* dummy for sscanf() */

      /* does the argument look like an address? */
      aflg = (4 == sscanf (argv[optind], "%d.%d.%d.%d", &i, &i, &i, &i));
    }
    if (aflg) {
	from_address(argv[optind], lflg) ;
    } else {
	from_name(argv[optind], lflg) ;
    }

    return 0 ;
}


from_address(astring, lflg)
char *astring ;
int lflg ;
{
    long addr ;			/* address in host order */
    long norder ;		/* address in network order */
    struct hostent *host ;	/* structure returned by gethostbyaddr() */
    int octet[4] ;		/* the four bytes of the address */

    /* parse string into address */
    if ((addr = inet_addr(astring)) == -1) {
	usage() ;
	exit(USAGE_ERROR) ;
    }

    /* try to find hostentry */
    if ((host = gethostbyaddr(&addr, 4, AF_INET)) == NULL) {
	gethost_error(progname, astring, 1) ;
    }

    /* print hostentry in long form or hostname */
    if (lflg) {
	print_long(host) ;
    } else {
	puts(host->h_name) ;
    }
}

from_name(name, lflg)
char *name ;
int lflg ;
{
    struct hostent *host ;

    /* try to find host entry */
    if ((host = gethostbyname(name)) == NULL) {
	gethost_error(progname, name, 0) ;
    }

    /* print host entry in long form or address */
    if (lflg) {
	print_long(host) ;
    } else {
	paddress(*host->h_addr_list) ;
    }
}

/* print host entry in long form */
print_long(host)
struct hostent *host ;
{
    int i ;
    unsigned long horder ;

    printf("Official name: %s\n", host->h_name) ;
    while (*host->h_aliases) {
	printf("Alias:         %s\n", *host->h_aliases++) ;
    }
    while (*host->h_addr_list) {
	printf("Address:       ") ;
	paddress(*host->h_addr_list++) ;
    }
}

/* print an internet address in network order with terminating newline */
paddress(paddr)
unsigned long *paddr ;
{
    unsigned long horder ;

    horder = ntohl(*paddr) ;
    printf("%u.%u.%u.%u\n", horder >> 24, (horder >> 16) & 0xff, 
	    (horder >> 8) & 0xff, horder & 0xff) ;

}
    

usage()
{
    fprintf(stderr, "\
Usage: %s [-l] [-n] hostname\n\
       %s [-l] [-a] IP-address\n\
-a forces interpretation of the argument as an address; -n forces\n\
interpretation of it as a hostname.  Otherwise, it is converted to\n\
a hostname if it looks like an address, and to an address if not.\n\
-l and -v both mean \"verbose output.\"\n",
	    progname, progname) ;
}

/* examine error condition */
gethost_error(s, name, addr_p)
char *s, *name ;
int addr_p;
{
    extern int h_errno ;	/* error code from gethostbyxxxx() */

    if (s && *s) {
	fputs(s, stderr) ;
	fputs(": ", stderr) ;
    }

    switch (h_errno) {
      case HOST_NOT_FOUND:
	fputs((addr_p ? "unknown address " : "unknown host "), stderr) ;
	fputs(name, stderr) ;
	fputs("\n", stderr) ;
	break ;
      case TRY_AGAIN:
	fputs("temporary error -- try again later\n", stderr) ;
	break ;
      case NO_RECOVERY:
	fputs("unrecoverable error\n", stderr) ;
	break ;
#ifdef NO_DATA
      case NO_DATA:
#else
#ifdef NO_ADDRESS
      case NO_ADDRESS:
#else
if neither NO_ADDRESS nor NO_DATA is defined, you are in trouble.
#endif /* NO_ADDRESS */
#endif /* NO_DATA  */
	fputs("no IP address\n", stderr) ;
	break ;
      default:
	fprintf(stderr, "unknown error number %d\n", h_errno) ;
	exit(UNKNOWN_ERROR) ;
    }
    exit(h_errno) ;
}
